### ISSUE TITLE
Make use of IDL to C++11 1.4 changed union member setters

### ISSUE THREAD

**Issue by jwillemsen** [OpenDDS/OpenDDS#3971]
See IDL to C++11 1.4 (https://issues.omg.org/issues/CPP1114-14), when there are multiple possible discriminator values the spec extends the API of the setter operations

**Comment by mitza-oci** [comment]
It's already implemented by opendds_idl.
An example is `tests/DCPS/Compiler/idl_test1_lib/FooDef.idl` which has union `Xyz::AUnion`.  The generated code for the `rsv()` and `usv()` setters have the optional 2nd parameter.

**Comment by jwillemsen** [comment]
Yes, you are correct, they are generated, my mistake.

I did notice they are not used when setting `rsv` in the generated code, there for rsv the discriminator and value are still set in 2 calls

**Comment by jwillemsen** [comment]
Concrete, it are `template<> void set_default( ::Xyz::AUnion& uni)` and `bool operator>>(Serializer& strm,  ::Xyz::AUnion& uni)`

**Comment by jwillemsen** [comment]
In `bool operator>>(Serializer& strm,  ::Xyz::AUnion& uni)` for the `gsv` member also 2 calls are done, but as there is only one discriminator for gsv the call to `_d(disc)` after setting the `gsv` member is not necessary

**Comment by mitza-oci** [comment]
Is it incorrect or just inefficient?

**Comment by jwillemsen** [comment]
Inefficient

**Comment by mitza-oci** [comment]
I don't think set_default is correct.  It uses `_d()` when the union is in an unknown state.

**Comment by ocielliottc** [comment]
I also don't think `set_default` is correct, but for a different reason.  Given the definition of `AUnion`, the default should be `Xyz::bluex`.  But, in `set_default`, it is set to `Xyz::redx`.  The constructor of `AUnion` sets the discriminator to `Xyz::bluex`.

**Comment by mitza-oci** [comment]
> Am I wrong or is the use of _d() to set the discriminator before setting the member value not only unnecessary, but also wrong?

In general, yes that's correct.  `_d()` can't be used to change the active member.  Once that's been attempted we're in to undefined behavior and it can do anything (including crash).

The default constructed union in the "classic" mapping has undefined state.  We could depend on the tao_idl implementation (I'm not sure what it uses to determine the default state).

The IDL-to-C++11 mapping has a definition of the default state.  But it may be simpler to not depend on this in the generated code (that way it works with either mapping).

**Comment by mitza-oci** [comment]
`set_default` uses the XTypes definitions of the default value (Table 9 of XTypes v1.3).  For this AUnion example the default discriminator is `redx` since it comes first in the enum.

**Comment by ocielliottc** [comment]
Ah, ok.  That makes things a little less complicated.

**Comment by jwillemsen** [comment]
Reopening, I see in the C++11/idl_test1_lib being generated in `bool operator>>(Serializer& strm,  ::Xyz::AUnion& uni)` the code

```
  switch (disc) {
  case  ::Xyz::ColorX::redx:
  case  ::Xyz::ColorX::yellowx: {
    std::string tmp;
    if (strm >> tmp) {
      uni.rsv(std::move(tmp));
      uni._d(disc);
      return true;
    }
```

But that should be

```
  switch (disc) {
  case  ::Xyz::ColorX::redx:
  case  ::Xyz::ColorX::yellowx: {
    std::string tmp;
    if (strm >> tmp) {
      uni.rsv(std::move(tmp)), disc;
      return true;
    }
```

In `template<> void set_default( ::Xyz::TestUnion3& uni)` the code can also be simplified

```
template<> void set_default( ::Xyz::TestUnion3& uni)
{
  ACE_UNUSED_ARG(uni);
  ::CORBA::Boolean temp;
  temp = 0;
  uni._d(temp);
}
```

to 

template<> void set_default( ::Xyz::TestUnion3& uni)
{
  ACE_UNUSED_ARG(uni);
  ::CORBA::Boolean temp;
  temp = 0;
  uni._d(false);
}

**Comment by mitza-oci** [comment]
Isn't the compiler going to produce the same code either way?  @ocielliottc what do you think about the 2-arg methods for `>>` and removing `temp` in `set_default`?

**Comment by ocielliottc** [comment]
It would seem that we should be able to do away with setting the discriminator in the `>>` methods.

As far as removing the temp, it isn't always that simple.  If the branch member is anything but a primitive, a temporary is going to be necessary.  I think it's fine as it is.  The compiler can optimize it away.

**Comment by mitza-oci** [comment]
Right, there would need to be some logic the code generator to determine:
* which types need `temp` (or similar) in `set_default`
* which branches of the union need the 2nd parameter in the setter in `>>`

**Comment by ocielliottc** [comment]
I'm not sure why we would need a two parameter setter.  Calling `uni.rsv(std::move(tmp))` would set the discriminator.  Right?

**Comment by mitza-oci** [comment]
It's for the branches where two or more distinct discriminator values are possible.

**Comment by ocielliottc** [comment]
Ah, yep.  I forgot about that.

**Comment by mitza-oci** [comment]
OK, we can leave this open if someone wants to change the generated code to something that's a little smaller and may be more efficient.
